File: frontend/src/services/workspaceService.ts
import { teamService } from './teamService';

export interface Workspace {
  id: string;
  name: string;
  owner_id: number;
  created_at: string;
  updated_at: string;
  plan: string;
  member_count: number;
  repository_count: number;
}

export interface WorkspaceMember {
  authProvider: any;
  id: number;
  workspace_id: string;
  user_id: number;
  name: string;
  email: string;
  avatar?: string;
  role: 'Owner' | 'Admin' | 'Member' | 'Viewer';
  status: 'Active' | 'Pending' | 'Inactive';
  invited_by?: number;
  joined_at: string;
  last_active?: string;
}

export interface Repository {
  id: number;
  github_id: number;
  name: string;
  full_name: string;
  description?: string;
  html_url: string;
  clone_url: string;
  language?: string;
  is_private: boolean;
  is_archived?: boolean;
  default_branch: string;
  created_at?: string;
  updated_at?: string;
}

class WorkspaceService {
  private readonly API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

  // Map team data to workspace format
  private mapTeamToWorkspace(team: any): Workspace {
    return {
      id: team.id.toString(),
      name: team.name,
      owner_id: team.created_by,
      created_at: team.created_at,
      updated_at: team.updated_at || team.created_at,
      plan: 'Pro Trial',
      member_count: 0,
      repository_count: 0,
    };
  }

  // Get all workspaces
  async getUserWorkspaces(): Promise<Workspace[]> {
    try {
      const token = localStorage.getItem('access_token');
      
      if (!token) {
        console.error('❌ No access token found');
        return [];
      }

      const response = await fetch(`${this.API_URL}/api/v1/workspace/list`, {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      const workspaces = data.workspaces || [];
      
      if (!Array.isArray(workspaces)) {
        console.error('❌ Workspaces is not an array:', workspaces);
        return [];
      }
      
      // Map to frontend format
      return workspaces.map((ws: any) => ({
        id: ws.id?.toString() || ws.team_id?.toString(),
        name: ws.name,
        owner_id: ws.owner_id,
        created_at: ws.created_at,
        updated_at: ws.updated_at,
        plan: ws.plan || 'Pro Trial',
        member_count: ws.member_count || 0,
        repository_count: ws.repository_count || 0,
      }));
      
    } catch (error) {
      console.error('❌ Error fetching workspaces:', error);
      return [];
    }
  }

  // Get workspace by ID
  async getWorkspace(workspaceId: string): Promise<Workspace> {
    const workspaces = await this.getUserWorkspaces();
    const workspace = workspaces.find(w => w.id === workspaceId);
    if (!workspace) {
      throw new Error('Workspace not found');
    }
    return workspace;
  }

  // Get workspace members (uses existing team members endpoint)
  async getWorkspaceMembers(workspaceId: string): Promise<WorkspaceMember[]> {
    try {
      const members = await teamService.getTeamMembers();
      return members.map(m => ({
        ...m,
        workspace_id: workspaceId,
        joined_at: m.dateJoined,
        last_active: m.lastActive || undefined,
      }));
    } catch (error) {
      console.error('Error fetching workspace members:', error);
      throw error;
    }
  }

  // Get workspace repositories
  async getWorkspaceRepositories(workspaceId: string): Promise<Repository[]> {
    try {
      const token = localStorage.getItem('access_token');
      
      if (!token) {
        throw new Error('No access token found');
      }

      const response = await fetch(`${this.API_URL}/api/v1/workspace/${workspaceId}/repositories`, {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to fetch workspace repositories');
      }

      return await response.json();
    } catch (error) {
      console.error('❌ Error in getWorkspaceRepositories:', error);
      throw error;
    }
  }

  // Get available repositories (not in workspace)
  async getAvailableRepositories(workspaceId: string): Promise<Repository[]> {
    try {
      const token = localStorage.getItem('access_token');
      
      if (!token) {
        throw new Error('No access token found');
      }

      const response = await fetch(
        `${this.API_URL}/api/v1/workspace/${workspaceId}/available-repositories`,
        {
          headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
          },
        }
      );

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to fetch available repositories');
      }

      return await response.json();
    } catch (error) {
      console.error('❌ Error in getAvailableRepositories:', error);
      throw error;
    }
  }

  // Add repository to workspace
  async addRepositoryToWorkspace(workspaceId: string, repositoryId: number): Promise<any> {
    try {
      const token = localStorage.getItem('access_token');
      
      if (!token) {
        throw new Error('No access token found');
      }

      const response = await fetch(
        `${this.API_URL}/api/v1/workspace/${workspaceId}/repositories`,
        {
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ repository_id: repositoryId }),
        }
      );

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to add repository');
      }

      return await response.json();
    } catch (error: any) {
      console.error('❌ Error in addRepositoryToWorkspace:', error);
      throw new Error(error.message || 'Failed to add repository to workspace');
    }
  }

  // Generate invite link (uses existing team endpoint)
  async generateInviteLink(workspaceId: string, role: string) {
    return await teamService.generateInviteLink(role);
  }

  // Send email invitations (uses existing team endpoint)
  async sendEmailInvitations(workspaceId: string, emails: string[], role: string) {
    return await teamService.sendEmailInvitations(emails, role);
  }

  // Update member role (uses existing team endpoint)
  async updateMemberRole(workspaceId: string, memberId: number, role: string) {
    return await teamService.updateMemberRole(memberId, role);
  }

  // Remove member (uses existing team endpoint)
  async removeMember(workspaceId: string, memberId: number) {
    return await teamService.removeMember(memberId);
  }

  // Switch workspace
  async switchWorkspace(workspaceId: string): Promise<void> {
    const token = localStorage.getItem('access_token');
    
    const response = await fetch(`${this.API_URL}/api/v1/workspace/switch`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ workspace_id: workspaceId }),
    });

    if (!response.ok) {
      throw new Error('Failed to switch workspace');
    }

    localStorage.setItem('current_workspace_id', workspaceId);
  }

  // Get current workspace ID
  getCurrentWorkspaceId(): string | null {
    return localStorage.getItem('current_workspace_id') || '1';
  }

  // Create workspace (for now, use GitHub OAuth flow)
  async createWorkspace(name: string): Promise<void> {
    const redirectUri = `${window.location.origin}/workspace/callback`;
    const state = encodeURIComponent(JSON.stringify({ name, action: 'create_workspace' }));
    
    sessionStorage.setItem('pending_workspace_name', name);
    
    window.location.href = `https://github.com/login/oauth/authorize?client_id=${import.meta.env.VITE_GITHUB_CLIENT_ID}&redirect_uri=${redirectUri}&scope=repo&state=${state}`;
  }

  // Delete workspace
  async deleteWorkspace(workspaceId: string): Promise<{ new_active_workspace_id: string }> {
    try {
      const token = localStorage.getItem('access_token');
      
      if (!token) {
        throw new Error('No access token found');
      }

      const response = await fetch(`${this.API_URL}/api/v1/workspace/${workspaceId}`, {
        method: 'DELETE',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to delete workspace');
      }

      const data = await response.json();
      
      return {
        new_active_workspace_id: data.new_active_workspace_id?.toString()
      };
      
    } catch (error) {
      console.error('❌ Error deleting workspace:', error);
      throw error;
    }
  }

  // Update workspace name
  async updateWorkspace(workspaceId: string, data: { name: string }): Promise<Workspace> {
    try {
      const token = localStorage.getItem('access_token');
      
      if (!token) {
        throw new Error('No access token found');
      }

      const response = await fetch(`${this.API_URL}/api/v1/workspace/${workspaceId}`, {
        method: 'PUT',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ name: data.name }),
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to update workspace');
      }

      const updatedWorkspace = await response.json();
      
      return {
        id: updatedWorkspace.id?.toString() || workspaceId,
        name: updatedWorkspace.name,
        owner_id: updatedWorkspace.owner_id,
        created_at: updatedWorkspace.created_at,
        updated_at: updatedWorkspace.updated_at,
        plan: updatedWorkspace.plan || 'Pro Trial',
        member_count: updatedWorkspace.member_count || 0,
        repository_count: updatedWorkspace.repository_count || 0,
      };
      
    } catch (error: any) {
      console.error('❌ Error updating workspace:', error);
      throw new Error(error.message || 'Failed to update workspace name');
    }
  }

  // ==========================================
  // GitHub PAT Token Management
  // ==========================================
  
  async checkGitHubTokenStatus(): Promise<any> {
    const token = localStorage.getItem('access_token');
    if (!token) throw new Error('No access token found');

    const response = await fetch(`${this.API_URL}/api/v1/github-integration/pat/status`, {
      headers: { 'Authorization': `Bearer ${token}` }
    });

    if (!response.ok) throw new Error('Failed to check token status');
    return response.json();
  }

  async saveGitHubToken(patToken: string): Promise<any> {
    const token = localStorage.getItem('access_token');
    if (!token) throw new Error('No access token found');

    const response = await fetch(`${this.API_URL}/api/v1/github-integration/pat/save`, {
      method: 'POST',
      headers: { 
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ token: patToken })
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to save token');
    }

    return response.json();
  }

  async deleteGitHubToken(): Promise<any> {
    const token = localStorage.getItem('access_token');
    if (!token) throw new Error('No access token found');

    const response = await fetch(`${this.API_URL}/api/v1/github-integration/pat/delete`, {
      method: 'DELETE',
      headers: { 'Authorization': `Bearer ${token}` }
    });

    if (!response.ok) throw new Error('Failed to delete token');
    return response.json();
  }

  // ==========================================
  // Utility functions
  // ==========================================

  validateEmail(email: string): boolean {
    return teamService.validateEmail(email);
  }

  parseEmails(emailString: string): string[] {
    return teamService.parseEmails(emailString);
  }

  validateEmails(emails: string[]) {
    return teamService.validateEmails(emails);
  }
}

export const workspaceService = new WorkspaceService();

File: backend/app/services/workspace_service.py
# backend/app/services/workspace_service.py
from sqlalchemy.orm import Session
from app.models.team import Team, TeamMember, MemberRole, MemberStatus
from app.models.user import User
from app.models.repository import Repository
from app.models.team_repository import TeamRepository
from app.services.github_service import GitHubService
from typing import List, Dict, Any, Optional
import logging
import traceback

logger = logging.getLogger(__name__)


class WorkspaceService:
    def __init__(self, db: Session):
        self.db = db
        self.github_service = GitHubService()

    def get_user_repositories(self, user: User) -> List[Dict[str, Any]]:
        """
        Fetch user's GitHub repositories using their stored token.
        """
        try:
            if not user. github_token:
                raise ValueError("User does not have a GitHub token. Please login with GitHub first.")
            
            repos = self.github_service.get_user_repositories(user.github_token)
            logger.info(f"Fetched {len(repos)} repositories for user {user.id}")
            return repos
        except Exception as e:
            logger.error(f"Error fetching repositories: {e}")
            raise

    def create_workspace_with_repositories(
        self,
        user_id: int,
        workspace_name: str,
        repository_ids: List[int],
        github_token: str
    ) -> Dict[str, Any]:
        """
        Create a new workspace and add selected repositories to it.
        """
        try:
            # Create new team/workspace
            team = Team(
                name=workspace_name,
                created_by=user_id
            )
            self.db.add(team)
            self.db.flush()
            logger.info(f"Created new team {team.id}")

            # Set as active workspace
            user = self.db.query(User).filter(User.id == user_id).first()
            user.active_team_id = team. id

            # Add creator as Owner in TeamMember
            team_member = TeamMember(
                team_id=team.id,
                user_id=user_id,
                role=MemberRole.owner,
                status=MemberStatus.active
            )
            self.db.add(team_member)
            logger. info(f"Added user {user_id} as Owner of team {team.id}")

            # Fetch repository details from GitHub
            all_repos = self.github_service.get_user_repositories(github_token)
            selected_repos = [r for r in all_repos if r['id'] in repository_ids]

            # Store repositories in database
            stored_repos = []
            for repo_data in selected_repos:
                # Check if repository already exists (by github_id and owner)
                existing_repo = self.db.query(Repository).filter(
                    Repository.github_id == repo_data['id'],
                    Repository. owner_id == user_id
                ).first()

                if existing_repo:
                    # Update existing repository
                    repo = existing_repo
                    repo.name = repo_data['name']
                    repo.full_name = repo_data['full_name']
                    repo.description = repo_data.get('description')
                    repo.html_url = repo_data['html_url']
                    repo.clone_url = repo_data['clone_url']
                    repo.language = repo_data.get('language')
                    repo.is_private = repo_data['private']
                    repo.default_branch = repo_data.get('default_branch', 'main')
                else:
                    # Create new repository
                    repo = Repository(
                        owner_id=user_id,
                        name=repo_data['name'],
                        full_name=repo_data['full_name'],
                        description=repo_data.get('description'),
                        html_url=repo_data['html_url'],
                        clone_url=repo_data['clone_url'],
                        language=repo_data.get('language'),
                        is_private=repo_data['private'],
                        default_branch=repo_data.get('default_branch', 'main'),
                        github_id=repo_data['id']
                    )
                    self.db.add(repo)
                    self.db.flush()
                
                stored_repos.append(repo)
                logger.info(f"Stored repository {repo.id} - {repo.name}")

            # Link repositories to team
            for repo in stored_repos:
                team_repo = TeamRepository(
                    team_id=team.id,
                    repository_id=repo.id
                )
                self.db.add(team_repo)
                logger.info(f"✅ Linked repository {repo. id} ({repo.name}) to team {team.id}")

            self.db.commit()
            
            return {
                'workspace_id': team.id,
                'workspace_name': team.name,
                'repository_count': len(stored_repos),
                'message': f'Workspace created with {len(stored_repos)} repositories'
            }
            
        except Exception as e: 
            self.db.rollback()
            logger.error(f"Error creating workspace: {e}")
            raise

    def get_user_workspaces(self, user_id: int) -> List[Dict[str, Any]]: 
        """
        Get all workspaces for a user. 
        """
        try: 
            # Get all teams where user is a member
            team_members = self.db.query(TeamMember).filter(
                TeamMember.user_id == user_id,
                TeamMember.status == MemberStatus.active
            ).all()
            
            workspaces = []
            for tm in team_members:
                team = tm.team
                
                # Count repositories
                repo_count = self.db.query(TeamRepository).filter(
                    TeamRepository.team_id == team. id
                ).count()
                
                # Count members
                member_count = self. db.query(TeamMember).filter(
                    TeamMember.team_id == team.id,
                    TeamMember. status == MemberStatus.active
                ).count()
                
                workspaces.append({
                    'id': team.id,
                    'name': team.name,
                    'owner_id': team.created_by,
                    'created_at': team.created_at.isoformat() if team.created_at else None,
                    'updated_at': team.updated_at.isoformat() if team.updated_at else None,
                    'repository_count': repo_count,
                    'member_count': member_count,
                    'plan': 'Pro Trial'
                })
            
            return workspaces
            
        except Exception as e:
            logger. error(f"Error getting user workspaces: {e}")
            raise

    def switch_workspace(self, user_id: int, workspace_id: int) -> Dict[str, Any]:
        """
        Switch active workspace for a user.
        """
        try:
            # Verify user has access to workspace
            member = self.db.query(TeamMember).filter(
                TeamMember.team_id == workspace_id,
                TeamMember.user_id == user_id,
                TeamMember.status == MemberStatus.active
            ).first()
            
            if not member: 
                raise ValueError("User is not a member of this workspace")
            
            # Update user's active workspace
            user = self.db. query(User).filter(User.id == user_id).first()
            user.active_team_id = workspace_id
            self.db.commit()
            
            return {
                'message':  'Workspace switched successfully',
                'workspace_id': workspace_id
            }
            
        except Exception as e:
            self.db.rollback()
            logger. error(f"Error switching workspace:  {e}")
            raise

    # ✅ Get workspace repositories
    def get_workspace_repositories(
        self,
        workspace_id: int,
        user_id: int
    ) -> List[Dict[str, Any]]:
        """
        Get all repositories associated with a workspace.
        """
        try:
            logger.info(f"🔍 Getting repositories for workspace {workspace_id}, user {user_id}")
            
            # Verify user has access to this workspace
            member = self.db.query(TeamMember).filter(
                TeamMember.team_id == workspace_id,
                TeamMember.user_id == user_id,
                TeamMember.status == MemberStatus.active
            ).first()
            
            if not member: 
                raise ValueError(f"User {user_id} is not a member of workspace {workspace_id}")
            
            # Get all repository IDs linked to this workspace
            team_repos = self.db.query(TeamRepository).filter(
                TeamRepository.team_id == workspace_id
            ).all()
            
            logger.info(f"📦 Found {len(team_repos)} team_repository entries")
            
            if not team_repos:
                logger.info(f"No repositories found for workspace {workspace_id}")
                return []
            
            # Get full repository details
            repo_ids = [tr.repository_id for tr in team_repos]
            repositories = self.db.query(Repository).filter(
                Repository.id.in_(repo_ids)
            ).all()
            
            logger.info(f"✅ Retrieved {len(repositories)} repository details")
            
            # Format response
            result = []
            for repo in repositories:
                result.append({
                    'id': repo.id,
                    'github_id': repo.github_id,
                    'name': repo.name,
                    'full_name': repo.full_name,
                    'description': repo.description,
                    'html_url': repo.html_url,
                    'clone_url': repo.clone_url,
                    'language': repo.language,
                    'is_private': repo.is_private,
                    'is_archived': repo.is_archived if hasattr(repo, 'is_archived') else False,
                    'default_branch': repo.default_branch,
                    'created_at':  repo.created_at.isoformat() if repo.created_at else None,
                    'updated_at': repo.updated_at.isoformat() if repo.updated_at else None
                })
            
            return result
            
        except ValueError: 
            raise
        except Exception as e: 
            logger.error(f"❌ Error fetching workspace repositories: {e}")
            logger.error(traceback.format_exc())
            raise Exception(f"Failed to fetch workspace repositories: {str(e)}")

    # ✅ Get available repositories (not in workspace)
    def get_available_repositories(self, user_id: int, workspace_id: int) -> List[Dict[str, Any]]: 
        """
        Get repositories that can be added to workspace (not already in it).
        """
        try:
            logger.info(f"🔍 Getting available repositories for user {user_id}, workspace {workspace_id}")
            
            # Get all user's repositories from database
            all_repos = self.db.query(Repository).filter(
                Repository.owner_id == user_id
            ).all()
            
            logger.info(f"📦 User has {len(all_repos)} total repositories")
            
            # Get repositories already in workspace
            workspace_repo_ids = self.db.query(TeamRepository. repository_id).filter(
                TeamRepository.team_id == workspace_id
            ).all()
            workspace_repo_ids = [r[0] for r in workspace_repo_ids]
            
            logger.info(f"📦 Workspace already has {len(workspace_repo_ids)} repositories")
            
            # Filter out repositories already in workspace
            available_repos = [
                {
                    'id': repo.id,
                    'github_id': repo.github_id,
                    'name': repo.name,
                    'full_name': repo.full_name,
                    'description': repo.description,
                    'language': repo.language,
                    'is_private': repo.is_private,
                    'html_url': repo.html_url
                }
                for repo in all_repos
                if repo.id not in workspace_repo_ids
            ]
            
            logger. info(f"✅ Found {len(available_repos)} available repositories")
            return available_repos
            
        except Exception as e:
            logger. error(f"❌ Error getting available repositories: {e}")
            logger.error(traceback.format_exc())
            raise

    # ✅ Add repository to workspace
    def add_repository_to_workspace(
        self,
        workspace_id: int,
        repository_id: int,
        user_id: int
    ) -> Dict[str, Any]:
        """
        Add an existing repository to a workspace.
        """
        try:
            logger.info(f"➕ Adding repository {repository_id} to workspace {workspace_id}")
            
            # Verify user is admin/owner of workspace
            member = self.db.query(TeamMember).filter(
                TeamMember.team_id == workspace_id,
                TeamMember.user_id == user_id,
                TeamMember.status == MemberStatus.active
            ).first()
            
            if not member or member.role. value not in ['Owner', 'Admin']:
                raise ValueError("Only workspace owners and admins can add repositories")
            
            # Check if repository exists and belongs to user
            repository = self.db.query(Repository).filter(
                Repository.id == repository_id,
                Repository.owner_id == user_id
            ).first()
            
            if not repository:
                raise ValueError("Repository not found or you don't have access")
            
            # Check if already linked
            existing_link = self.db.query(TeamRepository).filter(
                TeamRepository.team_id == workspace_id,
                TeamRepository.repository_id == repository_id
            ).first()
            
            if existing_link:
                raise ValueError("Repository is already in this workspace")
            
            # Create link
            team_repo = TeamRepository(
                team_id=workspace_id,
                repository_id=repository_id
            )
            self.db.add(team_repo)
            self.db.commit()
            
            logger.info(f"✅ Successfully added repository {repository_id} to workspace {workspace_id}")
            
            return {
                'message':  'Repository added successfully',
                'repository_id': repository_id,
                'workspace_id':  workspace_id
            }
            
        except ValueError: 
            self.db.rollback()
            raise
        except Exception as e: 
            self.db.rollback()
            logger.error(f"❌ Error adding repository to workspace: {e}")
            logger.error(traceback.format_exc())
            raise Exception(f"Failed to add repository:  {str(e)}")

    async def handle_workspace_oauth_callback(
        self,
        user_id: int,
        code: str,
        state: str
    ) -> Dict[str, Any]:
        """
        Handle OAuth callback for workspace creation.
        """
        # This would handle the OAuth flow
        # Implementation depends on your OAuth setup
        pass